feat: support `#__NO_SIDE_EFFECTS__` annotation for function declaration

antfu opened the pull request:
<!--
  ⚡️ katchow! We ❤️ Pull Requests!

  If you remove or skip this template, you'll make the 🐼 sad and the mighty god
  of Github will appear and pile-drive the close button from a great height
  while making animal noises.

  Pull Request Requirements:
  * Please include tests to illustrate the problem this PR resolves.
  * Please lint your changes by running `npm run lint` before creating a PR.
  * Please update the documentation in `/docs` where necessary

  Please place an x (no spaces - [x]) in all [ ] that apply.
-->

This PR contains:

- [ ] bugfix
- [x] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

### Description

Currently, the `#__PURE__` notation only supported on the function callee site, where it's usually functions' nature to be side-effect free. While Rollup already auto check if there would be side-effects inside a function to mark it pure, sometime the detection may not be perfect (https://github.com/rollup/rollup/issues/2962), especially for very complex functions.

This PR brings the capability of manually marking a function as pure on declarations, by introducing a new annotation syntax `#__NO_SIDE_EFFECTS__` along side `#__PURE__`. When using with a function declaration, it automatically makes all calls of the function side-effect free.

All the following examples work:

```ts
/*#__NO_SIDE_EFFECTS__*/
export function fnA (args) {
  // ...
}

export const fnB = /*#__NO_SIDE_EFFECTS__*/ (args) => {
  // ...
}

/*#__NO_SIDE_EFFECTS__*/ 
export const fnC = (args) => {
  // ...
}


/**
 * Some jsdocs
 *
 * @__NO_SIDE_EFFECTS__
 */
export const fnC = (args) => {
  // ...
}
```

### References

<!--
If this PR resolves any issues, list them as

  resolves #1234

where 1234 is the issue number. This will help us with house-keeping as Github will automatically add a note to those issues stating that a potential fix exists. Once the PR is merged, Github will automatically close those issues.

If an issue is only solved partially or is relevant in some other way, just list the number without "resolves".
-->

Existing Discussions:
- esbuild: https://github.com/evanw/esbuild/issues/1883
- UglifyJS: https://github.com/mishoo/UglifyJS/issues/5234

Existing workaround:
- https://github.com/danielroe/rollup-plugin-pure - injecting `/*#__PURE__*/` before callee
- https://github.com/fairysword/rollup-plugin-pure-annotation - the same

Solutions from other tools:
- https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#nosideeffects-modifies-thisarguments

Potentially solves:
- fix https://github.com/rollup/rollup/issues/2960
- fix https://github.com/rollup/rollup/issues/3334
- https://github.com/rollup/rollup/issues/2962 (requires manually adding the annotation)
- https://github.com/rollup/rollup/issues/2539 (requires manually adding the annotation)

References:
- my tweet: https://twitter.com/antfu7/status/1664320045797851147 

### TODO

- [ ] Documentation - I'll write them when the direction gets approved
- Use a more unambiguous annotation
  - https://twitter.com/lukastaegert/status/1664348230220492805
  - https://github.com/rollup/rollup/issues/3122
  - This can take some time to draft and discuss to find a good convention. Which this PR will focus on the logic handle using the existing __PURE__ syntax. It should be addressed in another PR.

Andarist commented:
As mentioned [here](https://github.com/mishoo/UglifyJS/issues/5234#issuecomment-1573232944) by @skyrpex it seems that there is an existing [`@nosideeffects`](https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#nosideeffects-modifies-thisarguments) annotation that perhaps would be a better choice here.

skyrpex commented:
> As mentioned [here](https://github.com/mishoo/UglifyJS/issues/5234#issuecomment-1573232944) by @skyrpex it seems that there is an existing [`@nosideeffects`](https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#nosideeffects-modifies-thisarguments) annotation that perhaps would be a better choice here.

Exactly. What we care about is about side effects, not whether the function is pure. Eg, a pure function wouldn't read global state, but if we don't use the data it returns and didn't cause any side effect, we can safely remove it.

lukastaegert commented:
> Which this PR will focus on the logic handle using the existing PURE syntax. It should be addressed in another PR.

While I appreciate the notion, I hope you will understand that I would not want to release this with the existing annotation. If we release this, then we do this to allow people to use it, and then we created precedence and basically changed the meaning of __PURE__ even if we did not want to.

> I'd be happy to adapt to google-closure-compiler's convention. Which also looks much more systematic. I am just a bit hesitant because then we will end up mixing with two different conventions for /*#__PURE__*/ and /*@nosideeffects*/ - maybe we should support @nosideeffects everywhere __PURE__ is supported?

I would actually be super happy to support existing annotations that are well-defined, regardless of style. If there are existing code-bases that use these annotations, they would immediately profit from it, and we can always extend annotation styles later.

In this case, however, `@nosideeffects` does not really work this way, as you can easily test yourself: https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520SIMPLE_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%2540formatting%2520pretty_print%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F%252F%2520ADD%2520YOUR%2520CODE%2520HERE%250A%252F**%2520%2540nosideeffects%2520*%252F%250Afunction%2520hello(name)%2520%257B%250A%2520%2520alert('Hello%252C%2520'%2520%252B%2520name)%253B%250A%257D%250A%250Ahello('New%2520user')%253B%250A%250A

To my limited knowledge of Closure Compiler, this annotation only has an effect in `@externs` files (with the corresponding header), which are like declaration files for external dependencies and can only contain empty declarations. I.e. those files would not be processed by Rollup anyway and Rollup would never see those annotations. On the other hand, supporting them in non-`@externs` files would be a change of meaning.

Here is my suggestion how to move forward without being blocked:
Just use a new annotation name that sounds nice without alignment across tools. Then we present the idea to all relevant tools, and either they say "fine" or come up with something better. In that case, we just support both annotations.

Taking a leaf out of Closure Compiler's book, `nosideeffects` is actually a better description than `pure` (i.e. always has the same output for same input AND has no side effects).

So how about
- `/*#__NO_SIDE_EFFECTS__*/`
- `/*#__NOSIDEEFFECTS__*/`
- `/*#__NOEFFECTS__*/`

For brevity, I would be leaning towards the latter, but you may have even better ideas.

lukastaegert commented on src/ast/nodes/shared/FunctionNode.ts:
```diff
@@ -45,6 +46,14 @@ export default class FunctionNode extends FunctionBase {
+
+		if (
+			(this.context.options.treeshake as NormalizedTreeshakingOptions).annotations &&
+			this.annotations
```
If we use a different annotation, of course it means we cannot use this check as there are different kinds of annotations with different effects.
Here would be one way of handling it:
- we still create `this.annotations` containing ALL annotations because we need this to remove all annotations of a Node if the Node is tree-shaken (it contains the exact locations of the comments we need to remove)
- These annotations are collected in `pureComments.ts`. To avoid re-reading all annotation comments, we add additional flags. Currently it works by adding a `_rollupAnnotations` property to the acorn Node, which is translated in `Node.ts` to the `annotations` property. How about depending on the annotation, we additionally add a `_rollupPure` or `_rollupNoSideEffects` property to the acorn Node that is copied over to the Rollup Nodes?
- We just need to make sure we also adapt the code in `CallExpression.ts` and `NewExpression.ts`.

antfu commented:
I have updated the implementation to use `#__NO_SIDE_EFFECTS__`.

- Renamed `pureComment.ts` to `commentAnnotations.ts`
- Add two new props to `NodeBase`, `annotationPure` and `annotationNoSideEffects`

The reason I chose it over `__NOSIDEEFFECTS__` is I consider it more readable. For `__NOEFFECTS__` it sounds good but I am not sure if "effects" over "side effects" could be too general. But in any case, I am open to changing it if there is a better option.

lukastaegert reviewed (commented):
Looks great, just some minor comments

lukastaegert reviewed (commented):
Looks great, just some minor comments

lukastaegert reviewed (commented):
Looks great, just some minor comments

lukastaegert reviewed (commented):
Looks great, just some minor comments

lukastaegert reviewed (approved):
Awesome, looks great! Will release this by tomorrow if there are no further comments.

antfu commented:
A side note, if you are using `esbuild` or `rollup-plugin-esbuild` to transpile your TypeScript file, the `/* @__NO_SIDE_EFFECTS__ */` comment might not be preserved (https://github.com/evanw/esbuild/issues/221).

A workaround to that is to change your comment block to `/*! @__NO_SIDE_EFFECTS__ */` (https://github.com/evanw/esbuild/commit/d7679fc7dff152daed0016d1aa7771f2d55e86eb). Not that would only work at the statement level.

Also, if you'd like to see this supported by esbuild natively, please vote or provide insights in this issue: https://github.com/evanw/esbuild/issues/3149
